### PR TITLE
Create a backup of the original app folder so we can still install the app

### LINKED ISSUE #14711: install an app if directory is still there.
When upgrading to a new owncloud version via the packages you can end up with the situation that an old app is still siting in the apps directory. This is deactivated correctly but a new and compatible apps can´t be installed from the app store because the old directory is still in the apps directory.
See here for the error:
https://www.bitblokes.de/2015/03/das-upgrade-auf-owncloud-8-0-1-hat-echt-nicht-alle-latten-am-zaun

If we detect during installation of an app that there is already a directory with the same name than we have to create an "legacy" directory (if it is not already there) and move the old directory into it.

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #14711 

Quick and dirty fix @DeepDiver1975 @karlitschek @PVince81 @georgehrke 
If we don't need the information in the UI:
issue-14711

Only the edit on `lib/private/installer.php` is required.

### CODE REVIEW

File: lib/private/app.php [hunk 1]
```diff
@@ -292,14 +293,14 @@ public static function enable($app, $groups = null) {
 	 * @param string $app
 	 * @return int
 	 */
-	public static function downloadApp($app) {
+	public static function downloadApp($app, array &$messages = array()) {
```
Comment (LukasReschke) [inline lib/private/app.php]: PHPDoc? :)

**Comment by jnfrmarks** [comment]
@nickvergessen 

What triggers the message?  We'll need that for testing.

**Comment by nickvergessen** [comment]
@jnfrmarks create a read-only folder `calendar` in the `apps/` folder and then try to install the calendar app via the apps page in your instance

File: lib/private/installer.php [hunk 2]
```diff
@@ -70,11 +70,26 @@ public static function installApp( $data = array()) {
...
+			$backupPath = '-backup';
+			$i = 0;
+			while (is_dir($basedir . $backupPath)) {
+				$i++;
+				$backupPath = '-backup-' . $i;
+			}
+
+			rename($basedir, $basedir . $backupPath);
+
+			if(is_dir($basedir)) {
```
Comment (PVince81) [inline lib/private/installer.php]: I guess this doesn't need a `clearstatcache()` ?

File: lib/private/app.php [hunk 3]
```diff
@@ -1058,11 +1059,12 @@ public static function getAppVersions() {
 
 	/**
 	 * @param mixed $app
+	 * @param array &$messages List of potential messages for the user
 	 * @return bool
 	 * @throws Exception if app is not compatible with this version of ownCloud
 	 * @throws Exception if no app-name was specified
 	 */
-	public static function installApp($app) {
+	public static function installApp($app, array &$messages = array()) {
```
Comment (PVince81) [inline lib/private/app.php]: I don't like pass by reference for messages but have no better idea that fits the time frame. So, fine.

**Comment by PVince81** [comment]
The code looks good.

File: lib/private/app.php (hunk 3 above)
Comment (nickvergessen) [inline lib/private/app.php]: Yeah, the only alternative way I see atm is not display a message at all and remove the new parameter everywhere

**Comment by PVince81** [comment]
I'm currently trying with a real calendar app instance, but it doesn't try to upgrade.
I guess I should try with the old "shipped" version ? Should the app be disabled before the attempt ? (probably disabled by the update)

**Comment by jnfrmarks** [comment]
@PVince81 

When going from 7 (where we ship calendar) to 8, the app is disabled as part of the upgrade process.

@nickvergessen 

Is that what a customer would do?  I want to make sure I have a good handle on the use case we're resolving so that I have the right test cases lined up.

**Comment by PVince81** [comment]
It is still not clear how a RPM/DEB package upgrade would have left the calendar app hanging instead of removing it during upgrade. But this case could possibly be reproduced by using tarballs instead and have the "apps" folder read-only.

**Comment by PVince81** [comment]
Nope. The update process DOES NOT disable the calendar/contact app because they are "shipped". This is the problem from https://github.com/owncloud/core/issues/14677 when you keep your apps folder as is.

I tried this from git:
1) checkout stable7 for core
2) checkout stable7 for the contacts app, which is inside "apps/contacts"
3) setup OC, see the contacts app is enabled
4) switch core to the branch from this PR
5) leave the contacts app on the old version (to simulate what we think happened for other people)
6) run the upgrade
7) :boom: contacts app not disabled, getting an exception because the old app is not compatible
(the problem from https://github.com/owncloud/core/issues/14677)

**Comment by PVince81** [comment]
Now I wonder if this is about upgrading from 8.0.0 to 8.0.1 instead, like this guy: https://github.com/owncloud/core/issues/14708 (not 7.0.4 to 8.0.1).

Because at this stage, all RPMs are already separated into separate packages, which means that the contacts/calendar apps are NOT shipped in this, so would be auto-disabled on upgrade.

I'll try this with git.

**Comment by PVince81** [comment]
Okay, based on my insights from https://github.com/owncloud/core/issues/14711#issuecomment-77413002 the way to go is testing 8.0.0 -> 8.0.1 upgrade

**Comment by PVince81** [comment]
#### Testing with read-write apps folder
1. Checkout v8.0.0 on core
2. Setup OC
3. Enable "contacts" and "calendar" from the "Recommended" section
4. Check "apps" folder and see that the apps were installed there
5. Checkout v8.0.1 on core
6. Run upgrade
7. Go to apps page, try enabling "contacts" or "calendar" from the "Recommended" section.
8. A message appear "app folder already exists" (not sure why the button says "Enable" instead of "Upgrade", see https://github.com/owncloud/core/issues/14708#issuecomment-77416392)
9. Switch to the branch from this PR
10. Try enabling them again: it works!

Please note that I didn't set anything to read-only. The message "app folder already exists" appeared even with those permissions !

**Comment by PVince81** [comment]
Before we merge this we need to look into the why the button says "Enable" instead of "Update", could it be an OCSID issue ? Because it is more likely to be the core issues than a read-only folder.

@karlitschek @georgehrke can you double check ?

**Comment by PVince81** [comment]
If the button said "Update" it is likely that there is logic to replace the old folder with a new one. (read-only case might need to be put there too, but still...)

**Comment by PVince81** [comment]
Just debugged: the OCS version is the same, so it's not technically an update.
It's just re-enable. But it still insists to redownload...

The "Enable" logic should probably check if the version on disk is already the wanted one, and instead of redownloading, just enable what is there.

**Comment by LukasReschke** [comment]
Can confirm @PVince81's assumption. Also matches the screenshot from the blog post.

**Comment by georgehrke** [comment]
> Before we merge this we need to look into the why the button says "Enable" instead of "Update", could it be an OCSID issue ? Because it is more likely to be the core issues than a read-only folder.

The issue here is that a new app was enabled for the calendar app in ownCloud 8. This new app is called  Calendar for ownCloud 8 or something like that. The ocsid we added to the calendar app before the release points to a different app (an empty calendar app). So ownCloud doesn't know that there is an relation between the already installed calendar and the Calendar for ownCloud 8.0. That's why it says enable instead of update

**Comment by PVince81** [comment]
OK so we might still need this PR here too for this special situation ?

Another issue: the backup folder in "apps" will still be detected as an app. So calendar would be detected twice. Might need to move it somewhere else.

**Comment by nickvergessen** [comment]
I think this is no longer needed due to @LukasReschke work?
Or do we still need it because of the different ocsid?

(If that is the case, I'd say we can ignore the displaying of the message and make the patch simple)

**Pull request closed without merging**